### PR TITLE
Refactor admission webhook client code to a apiserver/pkg/util package

### PR DESCRIPTION
**Pull request by mbohlool**
As part of #67006 This refactoring enable us to share code between admission webhooks and CRD conversion webhooks.

@deads2k @lavalamp @sttts @kubernetes/sig-api-machinery-misc

### CODE REVIEW

File: staging/src/k8s.io/api/admissionregistration/v1beta1/webhook_config_interface.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+package v1beta1
+
+import (
+	"k8s.io/apimachinery/pkg/util/json"
+)
+
+// Implementing k8s.io/apiserver/pkg/util/webhook.ClientConfigInterface
+
+func (c *WebhookClientConfig) GetURL() *string {
```
Comment (sttts) [inline staging/src/k8s.io/api/admissionregistration/v1beta1/webhook_config_interface.go]: see comments on conversion PR: this belongs where the interface is used. This is not API and should not live on the API types.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/initializer/initializer.go [hunk 2]
```diff
@@ -17,16 +17,15 @@ limitations under the License.
 package initializer
 
 import (
-	"net/url"
-
 	"k8s.io/apiserver/pkg/admission"
-	webhookconfig "k8s.io/apiserver/pkg/admission/plugin/webhook/config"
+	"k8s.io/client-go/util/webhook"
+	"net/url"
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/initializer/initializer.go]: import order

**Comment by sttts** [comment]
One structural comment, otherwise it looks good.

File: staging/src/k8s.io/client-go/util/webhook/client.go [hunk 3]
```diff
@@ -14,33 +14,37 @@ See the License for the specific language governing permissions and
...
 const (
 	defaultCacheSize = 200
 )
 
-var (
-	ErrNeedServiceOrURL = errors.New("webhook configuration must have either service or URL")
-)
+// ClientConfigInterface is the interface to access webhook client config to share webhook connect code between
+// admission webhooks and CRD conversion webhooks.
+type ClientConfigInterface interface {
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/util/webhook/client.go]: Why not an options struct instead of making folks implement an interface?

Comment (sttts) [reply staging/src/k8s.io/client-go/util/webhook/client.go]: +1. Matches our Go style much better.

Comment (mbohlool) [reply staging/src/k8s.io/client-go/util/webhook/client.go]: Interface was your suggestion Daniel :)

By option struct you mean an struct that has all of these values and constructor for any API that has those fields? I don't see the benefit of that only duplicating data.

Comment (mbohlool) [reply staging/src/k8s.io/client-go/util/webhook/client.go]: I made it a struct.

**Comment by mbohlool** [comment]
@lavalamp @sttts PTAL

**Comment by sttts** [comment]
lgtm and approved when it's green.

**Comment by mbohlool** [comment]
@sttts @lavalamp all green. pTAL

**Comment by caesarxuchao** [comment]
@mbohlool is this package going to be used by apiserver and apiextension-apiserver? If so it belongs to the apiserver repo instead of client-go, as this package doesn't help users writing controllers.

**Comment by caesarxuchao** [comment]
I took a look at #67006, indeed `"k8s.io/client-go/util/webhook"` is only used by apiserver and apiextension-apiserver. Why not leaving this utility package in the apiserver repo? apiextension-apiserver is importing apiserver anyway, it's not going to cause import cycle.

**Comment by mbohlool** [comment]
@deads2k suggested we move this to client-go. David, wdyt?

**Comment by deads2k** [comment]
> @deads2k suggested we move this to client-go. David, wdyt?

I just didnt' want it in apimachinery (thought I saw it there at some point).  putting in apiserver is fine with me.

**Comment by mbohlool** [comment]
Moved the webhook utility package back to apiserver :)

**Comment by sttts** [comment]
Overriding approval as only BUILD file in the root is changed.

**Pull request merged**